laser_filters in rolling fails to build on rhel targets

The package laser_filters in rolling has been detected as not building on all platforms in the buildfile rhel. The release manager for rolling will consider disabling this build if it continues to fail to build.
- jenkins_urls:
  - [rhel 8 x86_64](https://build.ros2.org/view/Rbin_rhel_el864/job/Rbin_rhel_el864__laser_filters__rhel_8_x86_64__binary)
This is being filed because this package is about to be blacklisted. If this ticket is resolved please review whether it can be removed from the blacklist that should cross reference here.

```
DEBUG: [ 71%] Linking CXX executable scan_to_cloud_filter_chain
DEBUG: /usr/bin/cmake -E cmake_link_script CMakeFiles/scan_to_cloud_filter_chain.dir/link.txt --verbose=1
DEBUG: /usr/bin/c++ -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld CMakeFiles/scan_to_cloud_filter_chain.dir/src/scan_to_cloud_filter_chain.cpp.o -o scan_to_cloud_filter_chain  -Wl,-rpath,/builddir/build/BUILD/ros-rolling-laser-filters-2.0.1/obj-x86_64-redhat-linux-gnu:/opt/ros/rolling/lib:/opt/ros/rolling/lib64: liblaser_scan_filters.so /opt/ros/rolling/lib/libstatic_transform_broadcaster_node.so /opt/ros/rolling/lib/liblaser_geometry.so /opt/ros/rolling/lib/librclcpp_lifecycle.so /usr/lib64/libtinyxml2.so /opt/ros/rolling/lib/libtf2_ros.so /opt/ros/rolling/lib/libmessage_filters.so /opt/ros/rolling/lib/librclcpp_action.so /opt/ros/rolling/lib/librcl_action.so /opt/ros/rolling/lib/libtf2_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libtf2_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libtf2_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libtf2_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/libtf2_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/libtf2_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/libaction_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libaction_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libaction_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libaction_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/libaction_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/libaction_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/libunique_identifier_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libunique_identifier_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libunique_identifier_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libunique_identifier_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/libunique_identifier_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/libunique_identifier_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/libcomponent_manager.so /opt/ros/rolling/lib/libclass_loader.so /opt/ros/rolling/lib/libcomposition_interfaces__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libcomposition_interfaces__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libcomposition_interfaces__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libcomposition_interfaces__rosidl_generator_py.so /opt/ros/rolling/lib/libcomposition_interfaces__rosidl_generator_c.so /opt/ros/rolling/lib/libcomposition_interfaces__rosidl_typesupport_c.so /opt/ros/rolling/lib/libsensor_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libsensor_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libsensor_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libsensor_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/libsensor_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/libsensor_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/libtf2.so /opt/ros/rolling/lib/libgeometry_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libgeometry_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libgeometry_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libgeometry_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/libgeometry_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/libgeometry_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib64/libconsole_bridge.so.1.0 /opt/ros/rolling/lib/librclcpp.so -lpthread /opt/ros/rolling/lib/liblibstatistics_collector.so /opt/ros/rolling/lib/liblibstatistics_collector_test_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/liblibstatistics_collector_test_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/liblibstatistics_collector_test_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/liblibstatistics_collector_test_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/liblibstatistics_collector_test_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/liblibstatistics_collector_test_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/libstd_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libstd_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libstd_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libstd_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/libstd_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/libstd_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/librosgraph_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/librosgraph_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/librosgraph_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/librosgraph_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/librosgraph_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/librosgraph_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/libstatistics_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libstatistics_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libstatistics_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libstatistics_msgs__rosidl_generator_py.so /opt/ros/rolling/lib/libstatistics_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/libstatistics_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/librcl_lifecycle.so /opt/ros/rolling/lib/librcl.so /opt/ros/rolling/lib/librcl_yaml_param_parser.so /opt/ros/rolling/lib/libyaml.so /opt/ros/rolling/lib/librcl_interfaces__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/librcl_interfaces__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/librcl_interfaces__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/librcl_interfaces__rosidl_generator_py.so /opt/ros/rolling/lib/librcl_interfaces__rosidl_generator_c.so /opt/ros/rolling/lib/librcl_interfaces__rosidl_typesupport_c.so /opt/ros/rolling/lib/libbuiltin_interfaces__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/libbuiltin_interfaces__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/libbuiltin_interfaces__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/libbuiltin_interfaces__rosidl_generator_py.so /opt/ros/rolling/lib/libbuiltin_interfaces__rosidl_generator_c.so /opt/ros/rolling/lib/libbuiltin_interfaces__rosidl_typesupport_c.so /opt/ros/rolling/lib/librmw_implementation.so /opt/ros/rolling/lib/libament_index_cpp.so /opt/ros/rolling/lib/librmw.so /opt/ros/rolling/lib/librcl_logging_spdlog.so /opt/ros/rolling/lib/librcl_logging_interface.so /opt/ros/rolling/lib/libtracetools.so /opt/ros/rolling/lib/liblifecycle_msgs__rosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/liblifecycle_msgs__rosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/librosidl_typesupport_introspection_cpp.so /opt/ros/rolling/lib/librosidl_typesupport_introspection_c.so /opt/ros/rolling/lib/liblifecycle_msgs__rosidl_typesupport_cpp.so /opt/ros/rolling/lib/librosidl_typesupport_cpp.so /opt/ros/rolling/lib/liblifecycle_msgs__rosidl_generator_py.so /usr/lib64/libpython3.6m.so /opt/ros/rolling/lib/liblifecycle_msgs__rosidl_generator_c.so /opt/ros/rolling/lib/liblifecycle_msgs__rosidl_typesupport_c.so /opt/ros/rolling/lib/librosidl_typesupport_c.so /opt/ros/rolling/lib/librcpputils.so /opt/ros/rolling/lib/librosidl_runtime_c.so /opt/ros/rolling/lib/librcutils.so -ldl
DEBUG: liblaser_scan_filters.so: undefined reference to `boost::system::generic_category()'
DEBUG: collect2: error: ld returned 1 exit status
DEBUG: make[2]: Leaving directory '/builddir/build/BUILD/ros-rolling-laser-filters-2.0.1/obj-x86_64-redhat-linux-gnu'
DEBUG: make[1]: Leaving directory '/builddir/build/BUILD/ros-rolling-laser-filters-2.0.1/obj-x86_64-redhat-linux-gnu'
DEBUG: make[2]: *** [CMakeFiles/scan_to_cloud_filter_chain.dir/build.make:211: scan_to_cloud_filter_chain] Error 1
DEBUG: make[1]: *** [CMakeFiles/Makefile2:219: CMakeFiles/scan_to_cloud_filter_chain.dir/all] Error 2
DEBUG: make: *** [Makefile:149: all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -42,7 +42,7 @@ if(BUILD_TESTING)
     ament_target_dependencies(${TEST_NAME} filters pluginlib rclcpp sensor_msgs)
     ament_add_test(
         ${TEST_NAME}
-        COMMAND 
+        COMMAND
             $<TARGET_FILE:${TEST_NAME}>
             --ros-args --params-file ${PROJECT_SOURCE_DIR}/test/${TEST_NAME}.yaml
             --gtest_output=xml:${RESULT_FILENAME}

--- a/include/laser_filters/array_filter.h
+++ b/include/laser_filters/array_filter.h
@@ -32,21 +32,19 @@
 
 #include <map>
 #include <iostream>
+#include <mutex>
 #include <sstream>
 
-#include "boost/thread/mutex.hpp"
-#include "boost/scoped_ptr.hpp"
 #include <sensor_msgs/msg/laser_scan.hpp>
 
 #include "filters/median.hpp"
 #include "filters/mean.hpp"
 #include "filters/filter_chain.hpp"
-#include "boost/thread/mutex.hpp"
 
 namespace laser_filters{
 
 /** \brief A class to provide median filtering of laser scans in time*/
-class LaserArrayFilter : public filters::FilterBase<sensor_msgs::msg::LaserScan> 
+class LaserArrayFilter : public filters::FilterBase<sensor_msgs::msg::LaserScan>
 {
 public:
   /** \brief Constructor
@@ -96,7 +94,7 @@ public:
       return false;
     }
 
-    boost::mutex::scoped_lock lock(data_lock);
+    std::lock_guard<std::mutex> lock(data_lock);
     scan_out = scan_in; ///Quickly pass through all data \todo don't copy data too
 
     if (scan_in.ranges.size() != num_ranges_) //Reallocating
@@ -122,7 +120,7 @@ private:
   rclcpp::Parameter range_config_;
   rclcpp::Parameter intensity_config_;
 
-  boost::mutex data_lock;                 /// Protection from multi threaded programs
+  std::mutex data_lock;                 /// Protection from multi threaded programs
   sensor_msgs::msg::LaserScan temp_scan_; /** \todo cache only shallow info not full scan */
 
   filters::MultiChannelFilterChain<float> *range_filter_;

--- a/include/laser_filters/median_filter.h
+++ b/include/laser_filters/median_filter.h
@@ -34,15 +34,11 @@
 #include <iostream>
 #include <sstream>
 
-#include "boost/thread/mutex.hpp"
-#include "boost/scoped_ptr.hpp"
-
 #include <sensor_msgs/msg/laser_scan.hpp>
 
 #include "filters/median.hpp"
 #include "filters/mean.hpp"
 #include "filters/filter_chain.hpp"
-#include "boost/thread/mutex.hpp"
 
 namespace laser_filters
 {
@@ -99,7 +95,7 @@ namespace laser_filters
         RCLCPP_ERROR(logging_interface_->get_logger(), "LaserMedianFilter not configured");
         return false;
       }
-      boost::mutex::scoped_lock lock(data_lock);
+      std::lock_guard<std::mutex> lock(data_lock);
       scan_out = scan_in; ///Quickly pass through all data \todo don't copy data too
 
       if (scan_in.ranges.size() != num_ranges_) //Reallocating
@@ -130,7 +126,7 @@ namespace laser_filters
     unsigned int filter_length_; ///How many scans to average over
     unsigned int num_ranges_;    /// How many data point are in each row
 
-    boost::mutex data_lock;                 /// Protection from multi threaded programs
+    std::mutex data_lock;                 /// Protection from multi threaded programs
     sensor_msgs::msg::LaserScan temp_scan_; /** \todo cache only shallow info not full scan */
 
     // rclcpp::parameter::ParameterVariant parameter_value_;

--- a/package.xml
+++ b/package.xml
@@ -13,17 +13,17 @@
 
   <build_depend>ament_cmake_auto</build_depend>
 
-  <depend>pluginlib</depend>
-  <depend>sensor_msgs</depend>
-  <depend>rclcpp</depend>
-  <depend>tf2</depend>
-  <depend>filters</depend>
-  <depend>message_filters</depend>
-  <depend>tf2_ros</depend>
-  <depend>laser_geometry</depend>
   <depend>angles</depend>
+  <depend>filters</depend>
+  <depend>laser_geometry</depend>
+  <depend>message_filters</depend>
+  <depend>pluginlib</depend>
+  <depend>rclcpp</depend>
   <depend>rclcpp_lifecycle</depend>
-  
+  <depend>sensor_msgs</depend>
+  <depend>tf2</depend>
+  <depend>tf2_ros</depend>
+
   <test_depend>ament_cmake_gtest</test_depend>
 
   <export>

